services/debugger/cache/embeddedcache/embeddedCache.go

```diff
@@ -0,0 +1,192 @@
+package embeddedcache
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path"
+	"sync"
+	"time"
+
+	"github.com/dgraph-io/badger/v3"
+	"github.com/dgraph-io/badger/v3/options"
+	"github.com/rudderlabs/rudder-server/config"
+	"github.com/rudderlabs/rudder-server/rruntime"
+	"github.com/rudderlabs/rudder-server/utils/logger"
+	"github.com/rudderlabs/rudder-server/utils/misc"
+)
+
+/*
+loadCacheConfig sets the properties of the cache after reading it from the config file.
+This gives a feature of hot readability as well.
+*/
+func (e *EmbeddedCache[E]) loadCacheConfig() {
+	config.RegisterDurationConfigVariable(30, &e.CleanupFreq, true, time.Second, "LiveEvent.cache.clearFreq") // default clearFreq is 15 seconds
+	config.RegisterIntConfigVariable(100, &e.Limiter, true, 1, "LiveEvent.cache.limiter")
+	config.RegisterDurationConfigVariable(5, &e.Ticker, false, time.Minute, "LiveEvent.cache.GCTime")
+	config.RegisterDurationConfigVariable(15, &e.queryTimeout, false, time.Second, "LiveEvent.cache.queryTimeout")
+	config.RegisterIntConfigVariable(3, &e.retries, false, 1, "LiveEvent.cache.retries")
+}
+
+/*
+EmbeddedCache is an in-memory cache. Each key-value pair stored in this cache have a TTL and one goroutine removes the
+key-value pair form the cache which is older than TTL time.
+*/
+type EmbeddedCache[E any] struct {
+	Limiter      int
+	retries      int
+	path         string
+	Origin       string
+	done         chan struct{}
+	closed       chan struct{}
+	Ticker       time.Duration
+	queryTimeout time.Duration
+	CleanupFreq  time.Duration // TTL time on badgerDB
+	once         sync.Once
+	Db           *badger.DB
+	Logger       logger.Logger
+}
+
+type badgerLogger struct {
+	logger.Logger
+}
+
+func (badgerLogger) Errorf(format string, a ...interface{}) {
+	_, _ = fmt.Fprintf(os.Stderr, format, a...)
+}
+
+func (badgerLogger) Warningf(format string, a ...interface{}) {
+	_, _ = fmt.Fprintf(os.Stderr, format, a...)
+}
+
+// Update writes the entries into badger db with a TTL
+func (e *EmbeddedCache[E]) Update(key string, value E) {
+	e.Init()
+	txn := e.Db.NewTransaction(true)
+	res := e.Read(key)
+	if len(res) >= e.Limiter {
+		res = res[len(res)-e.Limiter+1:]
+	}
```

Comment:
What are we doing here exactly? Can this truncation create issues somehow? I'd like to understand more.

Reply:
This is basically to ensure that we don't store more than x amount of events into the db at any point in time. So I would just want to remove the oldest elements(typically the starting elements) and append the newer entries into the cache

Reply:
Makes sense. Can you please add a comment?

Reply:
Added a comment